text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,431 @@
+
+# Alternatives
+
+An alternative way to avoid infinitely iterating Ranges would be a reinterpretation of the user-provided parameters in the currently *infinite* cases. For example, the currently infinite `Range(0, 10, -1)` could be re-arranged to mean `Range(10, 0, -1)`. An argument against transforming Range(0, 10, -1) into Range(10, 0, -1) is that Range evaluates expressions using the provided parameters in their meaning of `min` and `max` or `start` and `end`. In conjunction with the step parameter there is a directionality (trajectory) which does not only control the numbers of the sequence but also the order within the sequence. Instead of `0, 1, .., 9` the Range would produce  `10, 9, .., 1` -- different numbers in a different order.
+Furthermore, having empty Ranges when using `Range` with calculated, non-literal bounds also has an *expressive* use (see CSV reader example in Motivation section) which the reinterpretation of the parameters to always return a non-empty Range would preclude.
+
+Another fair alternative is to leave everything as it is. Pony has been in fairly wide-spread use academically and to some degree even commercially, without anyone reporting the issues that for the author were the reason to start working on this RFC -- which was hanging (eventually running out of memory) Pony code. One could therefore speculate that the side effects and instances in which currently infinite Ranges are created are sufficiently documented and known by the majority of users of Pony.
+
+# Unresolved questions
+
+The function `.is_empty()` was added to the proposal as a counterpart to `.is_infinite()`. However, it would be fair to ask whether such an `.is_empty()` is really necessary since there already is a `.has_next()`. While for infinite loops it makes sense to have an `.is_infinite()` which informs that `.has_next()` will *remain* `true`, in the opposite case of empty Ranges, `.has_next()` will be `false` right away, and an `.is_empty()` is therefore somewhat redundant.
```

Comment:
The difference in semantic between `has_next` and `is_empty` is that `is_empty` describes a static property of the `Range` that you can know in advance before using it, while `has_next` is a dynamic property while you are executing the `Range`. The `is_empty` function allows to take decisions about a `Range`, for instance when its `start`, `end` or `inc` parameters are results of calculation, before using it. As such, I don't see them as redundant.

For instance
```pony
let r = Range(0, 10, 1)
// Before using the range
// r.is_empty() --> false
// r.has_next() --> true
for i in r do
  // When i == 10
  // r.is_empty() --> false
  // r.has_next() --> false
end
```